### DISCUSSION TITLE
Christoffel-consistent factorization and covariant derivatives

### DISCUSSION THREAD

**Discussion by jakewelde** [stack-of-tasks/pinocchio#1663] [General]
Hi there,

I have some followups to the question [#1548](https://github.com/stack-of-tasks/pinocchio/issues/1548) about whether the $C(q,v)$ computed by Pinocchio is consistent with the Christoffel symbols. Just to make sure, does your answer mean that $C(q,v) = \Gamma_{ijk} v^k$, where $\Gamma_{ijk}$ are the Christoffel symbols (of the first kind)? 

Furthermore, is this still the case for floating-body systems, in which case I am asking about the "generalized Christoffel symbols" in agreement with the implicit basis of vector fields used to express the components of the velocity $v$?

Finally, is this just the result of choosing a consistent body-level factorization, which therefore yields an appropriate factorization for the coupled system, as described in [this paper](https://arxiv.org/pdf/2010.01033.pdf) as a "Christoffel-consistent" factorization?

The reason I ask is that I am interested in using Pinocchio to compute covariant derivatives of various vector fields, so if the factorization used by Pinocchio is as I described, I believe that given any two vector fields represented in components in the basis of the velocity components, I can compute the covariant derivative of one along the other using a combination of matrix operations with $M$ and $C$ and Lie derivatives.

Many thanks,
Jake

**Comment by jcarpent** [comment]
Dear Jake,

Thanks for raising these questions.
The algorithm developed in Pinocchio for computing $C(q,v)$  has been developed together with the analytical derivatives of forward and inverse dynamics in 2017 and 2018 (https://hal.archives-ouvertes.fr/hal-01790971) and has been recently published in [the paper](https://arxiv.org/pdf/2010.01033.pdf) you mentioned.
After discussing with Patrick Wensing, we agree that the two algorithms were equivalent, so this answers your first question.

For the second one, in Pinocchio we always operate on the tangent space of the given motion groups enabled by the joints for velocity and forces quantities.

In Pinocchio, to get access to higher tensor derivatives, you can always rely on autodiff support, which is well supported by Pinocchio.
We have recently extended the support of Pinocchio with Python bindings accounting for Casadi and CppAD framework. This will be released soon.

Best,
Justin

**Comment by jakewelde** [comment]
Thanks Justin for your quick and helpful reply, it resolves my questions! Indeed, this question evolved out of a discussion with Dr. Wensing this week at ICRA. I’ll look further into implementing these things and I appreciate your insight.
Best,
Jake

**Comment by pwensing** [comment]
Hey guys,

As Justin says, the algorithm in my paper and in Pinnochio have the same conceptual operation. My algorithm and the supporting theory were developed in 2014, and I've shared documentation of that with Justin (I'm dreadfully slow to publish, and I'm sorry that was the case). I think we've agreed that there was independent development and value brought from both sides. I'd mostly been thinking about applying the Coriolis matrix for adaptive control, so I'm excited to see Jake's new directions here.

There are a few nuance points that I think will be important for Jake.

1. **The current implementation in Pinnochio gives a Corioils matrix that satisfies all the properties needed for most applications, but I believe it does not give the Coriolis matrix that comes from the Chirstoffel symbols**:  Note that Line 6 in Algorithm 1 of my paper ([link](https://doi.org/10.1115/1.4051169), [arxiv](https://arxiv.org/pdf/2010.01033.pdf)) uses $\mathbf{B} = \frac{1}{2}[(\mathbf{v} {\times^\ast})\mathbf{I}  + (\mathbf{I}\mathbf{v} \overline{\times}{{}^\ast}) - \mathbf{I} (\mathbf{v} \times)]$. This matrix is one choice satisfying $\mathbf{B}\mathbf{v} = \mathbf{v} \times^\ast \mathbf{I} \mathbf{v}$, and it is this particular choice that makes the algorithm give the Christoffel-consistent factorization. Propositions 1 and 2 in the paper discuss the relationship between the choice of $\mathbf{B}$ and the properties of the resulting Coriolis matrix. Pinnochio uses $\mathbf{B} = \mathbf{v} \times^\ast \mathbf{I}$ on [Line 453](https://github.com/stack-of-tasks/pinocchio/blob/380e4167e38307874c05b774c6a78b2e3575fd8a/src/algorithm/rnea.hxx#L453) of its algorithm. While this choice ensures that the resulting Coriolis matrix satifies $\dot{\mathbf{H}}- 2 \mathbf{C}$, the matrix is not the same one you get from the Christoffel symbols.
    *  @jcarpent, You probably recognize my formula for $\mathbf{B}$ from the RNEA derivatives. If you replace the line I linked above with half of the initial value of `doYcrb` from Lines [276](https://github.com/stack-of-tasks/pinocchio/blob/380e4167e38307874c05b774c6a78b2e3575fd8a/src/algorithm/rnea-derivatives.hxx#L276) and [278](https://github.com/stack-of-tasks/pinocchio/blob/380e4167e38307874c05b774c6a78b2e3575fd8a/src/algorithm/rnea-derivatives.hxx#L278) of [rnea-derivatives.hxx](https://github.com/stack-of-tasks/pinocchio/blob/380e4167e38307874c05b774c6a78b2e3575fd8a/src/algorithm/rnea-derivatives.hxx), then you'd get Jake what he is looking for.
    * @jakewelde, Until Justin makes that change, you can find my MATLAB implementation here ([link](https://github.com/ROAM-Lab-ND/spatial_v2_extended/blob/main/v3/dynamics/CoriolisMatrix.m)), though, of course, Pinnochio is the way to go for anything real time.
2. **There are a few subtleties when it comes to working with generalized speeds that should be noted**: If we have some basis vector fields $X_1, \ldots, X_n$ on the configuration manifold $\mathcal{Q}$, then the generalized Christoffel symbols (of the second kind) are given by $\nabla_{X_j} X_k = \sum_i \Gamma^i_{jk} X_i$ where $\nabla$ denotes the Riemannian connection. Similar to usual, we have the symbols of the first kind via $\Gamma_{ijk} = \sum_\ell H_{i\ell} \Gamma^\ell_{jk}$. In the usual case when $X_1,\ldots,X_n$ are coordinate vector fields, then $\Gamma_{ijk}$ is symmetric in the indices $jk$ (i.e., $\Gamma_{ijk}= \Gamma_{ikj}$). However, if your basis vector fields don't commute, you lose this property. This happens, for example, when working with floating-base joints. If Justin makes the change above (or if you use my MATLAB library), then the Coriolis matrix you get will satisfy $C_{ij} = \sum_k \Gamma_{ikj} v^k$. When you have a robot with all revolute/prismatic joints, then this is the same as $C_{ij} = \sum_k \Gamma_{ijk} v^k$ due to the symmetry of the symbols, but that doesn't hold in general.
    * @jakewelde, Proposition 2 of my paper only discusses the case when you have all revolute/prismatic joints, but I've verified after our ICRA conversation that the algo works even with more general joints (e.g., floating base) in the sense that you get $C_{ij} = \sum_k \Gamma_{ikj} v^k$. The root reason everything works out is that Theorem 1 in the paper generalizes to the case of transformations of generalized speeds.

Sorry for the long post, but I figured it was worth clarifying -- happy to follow up in any way that would be helpful.

Cheers,
-Pat

**Reply by jcarpent** [reply]
Thanks @pwensing for your precisions.
One quick and simple solution could consist in providing an extra signature to compute the Coriolis matrix with consistent Christoffel symbols. The current signature is pretty useful for computing quantities which are useful for other extra derivatives in fact.

**Reply by jakewelde** [reply]
I see - I may experiment with that on my end then! Thank you!

**Reply by pwensing** [reply]
Hey guys,

I recently put a preprint online ([link](https://arxiv.org/abs/2312.14425)), developing some of the comments above that I had put forward without proof. 

Big picture: the preprint explains how you can relate Coriolis factorizations in excess velocity coordinates with associated factorizations in minimal velocity coordinates. It also details what properties are inherited in this process. This, for instance, occurs when considering the dynamics of a constrained mechanism (e.g., when considering joints as constraints on a set of unconstrained rigid bodies). The $\mathbf{B}$ matrix above is intimately linked with the Riemannian connection for a single-rigid body, while the Coriolis algorithm is essentially projecting that information from maximal velocity coordinates to minimal ones. Overall, the key machinery for these developments is to relate Coriolis factorizations with affine connections (i.e., ways of taking covariant derivatives) on the configuration manifold, where the Christoffel-consistent factorization is the unique one associated with the Riemannian connection. 

Here is a quick set of pointers for those who are trying to understand the claims above:
- Proposition V.3 is most relevant to what I mentioned above as related to the Christoffel-consistent factorization for a constrained mechanism.
- Equations (18), (19), and (21) explain the link between the $\mathbf{B}$ matrix above and the Riemannian connection for a single rigid body.
- Section VI.B relates that theory to the Coriolis matrix algorithm for open-chain systems. 
- Section VI.C treats closed chains, which I think is relevant to some of Jake’s current work. 

I hope these pointers might be helpful if others stumble upon this thread down the road. I’m also happy to answer any questions that might come up. 

Happy New Year!

Best,
-Pat

**Comment by pwensing** [comment]
I do think this impacts the answer to question [#1548](https://github.com/stack-of-tasks/pinocchio/issues/1548) as well. It seems that the answer there assumes that the skew-symmetry of $\dot{\mathbf{H}} - 2 \mathbf{C}$ is enough to ensure that the Coriolis matrix is the same as the one coming from the Christoffel symbols, which is not the case.

**Comment by jakewelde** [comment]
Hi Pat,

No need to apologize for the long post, this was extremely helpful! I guess in hindsight you explained some of these things when we spoke verbally and also in the paper I linked, but I don't think I properly grasped them til now. For example - that the passivity property holds for a number of admissible factorizations, not just the Christoffel-consistent one. Thanks also for the line-by-line details about the differences between these choices in implementations, that's super helpful.

In regards to the non-commuting basis vector fields in general - this makes sense to me for the non-symmetry of the generalized Christoffel symbols, and I see you noted that the correct indexing $C_{ij} = \sum_k \Gamma_{ikj} v^k$ is different than my original question due to this subtlety. Just to confirm, this doesn't introduce any additional algorithmic problems in the actual algorithms for computing Coriolis matrix we're discussing here, right? It just needs to be accounted when I go to try to use this matrix to compute covariant derivatives, since I want to essentially extract the generalized Christoffels by evaluating $C$ on "one-hot" unit vectors for the velocity.

Justin, do you think Pinocchio is likely to have the option to use the body-factorization Pat suggests? I wonder if perhaps you chose the other one for performance reasons since it seems to be less total operations. Would having a flag to choose the factorization be an option? 

By the way, @Brian-Acosta [had a similar interest](https://github.com/RobotLocomotion/drake/issues/14315) regarding the passivity-based control that Pat was getting at. I discussed this briefly with Brian and I don't think he got as far as implementing it, but I think there's a few folks interested in being able to use this factorization. I'm not knowledgeable about passivity-based control, but perhaps the feature Brian was looking for is already available in Pinocchio (whereas he was originally asking about Drake) because the passivity property is satisfied, just not the Christoffel-consistency.

Thanks again to you both for your prompt and helpful messages.

Best,
Jake

**Reply by jcarpent** [reply]
> Justin, do you think Pinocchio is likely to have the option to use the body-factorization Pat suggests? I wonder if perhaps you chose the other one for performance reasons since it seems to be less total operations. Would having a flag to choose the factorization be an option?

I've just replied above ;)

**Reply by pwensing** [reply]
> Just to confirm, this doesn't introduce any additional algorithmic problems in the actual algorithms for computing the Coriolis matrix we're discussing here, right?

Correct.

> It just needs to be accounted when I go to try to use this matrix to compute covariant derivatives, since I want to essentially extract the generalized Christoffels by evaluating on "one-hot" unit vectors for the velocity.

Agreed :) 

Re: Brian's post: Thanks for the pointer! Of note is that $\mathbf{C} = \frac{1}{2} \frac{\partial}{\partial \dot{\mathbf{q}}} \mathbf{C}(\mathbf{q},\dot{\mathbf{q}}) \dot{\mathbf{q}}$  works just fine with generalized coordiantes (i.e., it gives you the Christoffel-consistent factorization), the result $\mathbf{C} = \frac{1}{2} \frac{\partial}{\partial \boldsymbol{v}} \mathbf{C}(\mathbf{q},\boldsymbol{v}) \boldsymbol{v}$ doesn't satisfy desired properties when working with generalized speeds (due to the same subtleties regarding non-commuting basis vector fields for the generalized speeds).

**Reply by jakewelde** [reply]
Ah, that's good to know! Thanks! :)

**Comment by jcarpent** [comment]
@jakewelde @pwensing I made the change (see https://github.com/stack-of-tasks/pinocchio/pull/1665) to get a Coriolis matrix which is consistent with Christoffel symbols of first kind.

@jakewelde You should be able to install Pinocchio from the devel branch from source.

**Reply by pwensing** [reply]
So fast! Thanks, Justin, for the quick attention.

**Reply by jakewelde** [reply]
Thanks so much Justin for your lightning-fast updates! Looking forward to making use of this.

**Reply by bhtxy0525** [reply]
@jcarpent Great！I can't wait to verify this new Coriolis matrix
